src/Cafe/OS/libs/nsyshid/Dimensions.cpp

```diff
@@ -0,0 +1,1076 @@
+		return true;
+	}
+
+	void DimensionsUSB::GenerateRandomNumber(uint8* buf, uint8 sequence,
+											 std::array<uint8, 32>& replyBuf)
+	{
+		// Decrypt payload into an 8 byte array
+		std::array<uint8, 8> value = Decrypt(buf, std::nullopt);
+		// Seed is the first 4 bytes (little endian) of the decrypted payload
+		uint32 seed = uint32(value[3]) << 24 | uint32(value[2]) << 16 | uint32(value[1]) << 8 | uint32(value[0]);
+		// Confirmation is the second 4 bytes (big endian) of the decrypted payload
+		uint32 conf = uint32(value[4]) << 24 | uint32(value[5]) << 16 | uint32(value[6]) << 8 | uint32(value[7]);
+		// Initialize rng using the seed from decrypted payload
+		InitializeRNG(seed);
+		// Encrypt 8 bytes, first 4 bytes is the decrypted confirmation from payload, 2nd 4 bytes are blank
+		std::array<uint8, 8> valueToEncrypt = {value[4], value[5], value[6], value[7], 0, 0, 0, 0};
+		std::array<uint8, 8> encrypted = Encrypt(valueToEncrypt.data(), std::nullopt);
+		replyBuf[0] = 0x55;
+		replyBuf[1] = 0x09;
+		replyBuf[2] = sequence;
+		// Copy encrypted value to response data
+		memcpy(&replyBuf[3], encrypted.data(), encrypted.size());
+		replyBuf[11] = GenerateChecksum(replyBuf, 11);
+	}
+
+	void DimensionsUSB::GetChallengeResponse(uint8* buf, uint8 sequence,
+											 std::array<uint8, 32>& replyBuf)
+	{
+		// Decrypt payload into an 8 byte array
+		std::array<uint8, 8> value = Decrypt(buf, std::nullopt);
+		// Confirmation is the first 4 bytes of the decrypted payload
+		uint32 conf = uint32(value[0]) << 24 | uint32(value[1]) << 16 | uint32(value[2]) << 8 | uint32(value[3]);
+		// Generate next random number based on RNG
+		uint32 nextRandom = GetNext();
+		// Encrypt an 8 byte array, first 4 bytes are the next random number (little endian)
+		// followed by the confirmation from the decrypted payload
+		std::array<uint8, 8> valueToEncrypt = {uint8(nextRandom & 0xFF), uint8((nextRandom >> 8) & 0xFF),
+											   uint8((nextRandom >> 16) & 0xFF), uint8((nextRandom >> 24) & 0xFF),
+											   value[0], value[1], value[2], value[3]};
+		std::array<uint8, 8> encrypted = Encrypt(valueToEncrypt.data(), std::nullopt);
+		replyBuf[0] = 0x55;
+		replyBuf[1] = 0x09;
+		replyBuf[2] = sequence;
+		// Copy encrypted value to response data
+		memcpy(&replyBuf[3], encrypted.data(), encrypted.size());
+		replyBuf[11] = GenerateChecksum(replyBuf, 11);
+	}
+
+	void DimensionsUSB::InitializeRNG(uint32 seed)
+	{
+		m_randomA = 0xF1EA5EED;
+		m_randomB = seed;
+		m_randomC = seed;
+		m_randomD = seed;
+
+		for (int i = 0; i < 42; i++)
+		{
+			GetNext();
+		}
+	}
+
+	uint32 DimensionsUSB::GetNext()
+	{
+		uint32 e = m_randomA - std::rotl(m_randomB, 21);
+		m_randomA = m_randomB ^ std::rotl(m_randomC, 19);
+		m_randomB = m_randomC + std::rotl(m_randomD, 6);
+		m_randomC = m_randomD + e;
+		m_randomD = e + m_randomA;
+		return m_randomD;
+	}
+
+	std::array<uint8, 8> DimensionsUSB::Decrypt(const uint8* buf, std::optional<std::array<uint8, 16>> key)
```

Comment:
This function blindly indexes a pointer and assumes it's an array that has at least 8 elements. This is fine as long as you validate that this function is only called with that precondition satisfied.
I know I risk this review being nothing but "use span" spam, but changing ``const uint8* buf`` to ``std::span<const uint8, 8>`` makes it harder to call the function with a buffer pointing to invalid memory in the first place, so you're less likely to make a mistake.
same goes for DimensionsUSB::Encrypt

Reply:
Same as above, decrypt is called sometimes from requests from the game (where the message is a pointer) which is why the method takes in a pointer. If there is a way to convert the pointer to a span of size originalLength as above, then this can change the rest of the methods to allow for a span param